src/backends/perf_pt/decode.c

```diff
@@ -35,13 +35,317 @@
+ *
+ * Returns true on success or false otherwise.
+ */
+static bool
+load_self_image(struct load_self_image_args *args)
+{
+    bool ret = dl_iterate_phdr(load_self_image_cb, args) == 0;
+    if (fsync(args->vdso_fd) == -1) {
+        ret = false;
+    }
+    return ret;
+}
+
+/*
+ * The callback for `load_self_image()`, called once for each program header.
+ *
+ * Returns 1 to stop iterating, and in our case to indicate an error. Returns 0
+ * on success and to continue iterating. See dl_iterate_phdr(3) for information
+ * on this interface.
+ */
+static int
+load_self_image_cb(struct dl_phdr_info *info, size_t size, void *data)
+{
+    ElfW(Phdr) phdr;
+    ElfW(Half) i;
+
+    (void) size;
+    struct load_self_image_args *args = data;
+
+    const char *filename = info->dlpi_name;
+    bool vdso = false;
+    if (!*filename) {
+        // On Linux, an empty name means that it is the executable itself.
+        filename = program_invocation_name;
+    } else {
+        vdso = strcmp(filename, VDSO_NAME) == 0;
+    }
+
+    for (i = 0; i < info->dlpi_phnum; i++) {
+        phdr = info->dlpi_phdr[i];
+        // Ensure we only use loadable and executable sections.
+        if ((phdr.p_type != PT_LOAD) || (!(phdr.p_flags & PF_X))) {
+            continue;
+        }
+
+        uint64_t vaddr = info->dlpi_addr + phdr.p_vaddr;
+        uint64_t offset;
+
+        // Load the code into the libipt image.
+        //
+        // The VDSO is special. It doesn't exist on-disk as a regular library,
+        // but rather it is a set of pages shared with the kernel.
+        //
+        // XXX Since libipt currently requires us to load from a file, we have
+        // to dump the VDSO to disk and have libipt load it back in.
+        //
+        // Discussion on adding libipt support for loading from memory here:
+        // https://github.com/01org/processor-trace/issues/37
+        if (vdso) {
+            int rv = dump_vdso(args->vdso_fd, vaddr, phdr.p_filesz);
+            if (!rv) {
+                return 1;
+            }
+            filename = args->vdso_filename;
+            offset = 0;
+        } else {
+            offset = phdr.p_offset;
+        }
+
+        // XXX This could be made faster using a libipt instruction cache.
+        int rv = pt_image_add_file(args->image, filename, offset,
+                                   phdr.p_filesz, NULL, vaddr);
+        if (rv < 0) {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+/*
+ * Dump the VDSO code into the open file descriptor `fd`, starting at `vaddr`
+ * and of size `len` into a temp file.
+ *
+ * Returns true on success or false otherwise.
+ */
+bool
+dump_vdso(int fd, uint64_t vaddr, size_t len)
+{
+    size_t wrote_so_far = 0;
+    while (wrote_so_far != len) {
+        int wrote = write(fd, (void *) vaddr + wrote_so_far, len - wrote_so_far);
+        if (wrote == -1) {
+            return false;
+        }
+        wrote_so_far += wrote;
+    }
+
```

Comment:
I think all of the other functions up to this point haven't had a blank line before `return`.

Reply:
Just ha d a look. It's currently a mix.